Compile external scripts off the main thread

AbhishekSharma102 opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #26571  (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @ferjm (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/globalscope.rs
 * @kichjang: components/script/dom/globalscope.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

AbhishekSharma102 commented:
Hey @gterzian I have some questions regarding this.

1. FinishOffThread should be called on the main thread or the callback? Documentation says on the runtime's main thread. If that's the case, should we store it in the callbackData and call it in evaluate_script_on_global_with_result?

2. evaluate_script_on_global_with_result should have the CompileOffThread call or run_a_classic_script? 

3. Could you please explain more on networking_task_source?

Sorry if some questions seem trivial.

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -2218,6 +2226,16 @@ impl GlobalScope {
+        callbackData: *mut c_void,
+    ) {
+        let castedCallbackData = callbackData as *mut OffThreadCompilationData;
+        let script: JSScript* = FinishOffThreadScript(*(castedCallbackData.context), token);
```
To answer your question: here you would use the `OffThreadCompilationData` to then queue a task, and inside that task you would do the rest such as calling `FinishOffThreadScript`. 

Also, the context can be obtained from inside the task, by calling `global.get_cx()` and then you need to de-reference it when needed, for example by doing `FinishOffThreadScript(*global.get_cx(), token);`

(and yes `context` is often used in the other structs I pointed to to refer to the `Trusted<GlobalScope>` but that is just a name to say it's the "context" for the overal task that is being queued. 

So I think this should be enough info to queue the task, and then let me know if you have other questions as to how to handle the `JsScript` further, I would say you should look into calling a method on the global with it.

AbhishekSharma102 replied:
Like other structs this would also not have access to the original context, and needs to be stored. Is that correct?

gterzian replied:
Yes, correct, you need to store the `Trusted<GlobalScope>`(and other things if you need them.

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -285,6 +285,14 @@ struct FileListener {
     task_canceller: TaskCanceller,
 }
 
+#[derive(JSTraceable, MallocSizeOf)]
```
Since this struct will be used in the IPC router only, you don't need to derive those traits which are mostly meant for stuff inside a struct marked as `dom_struct`.

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -285,6 +285,14 @@ struct FileListener {
+pub struct OffThreadCompilationData {
+    context: JSContext
+    canceller: TaskCanceller,
+    task_source: TimerTaskSource,
```
To answer your other question, so here you should instead store the `NetworkingTaskSource`, which is available via a call to `GlobalScope.networking_task_source()`. 

Then later when you need to queue a task, you can do `OffThreadCompilationData.task_source.queue_with_canceller`, like is done for example at https://github.com/servo/servo/blob/d7d56454b07b70e3d9b2e3746ea1b2a1a1f05b7f/components/script/dom/globalscope.rs#L404

gterzian replied:
I'd say you can define a `handle(&self, token: Token)` method on the struct, and then from the callback call `handle(token)`, and inside the method queue the task and move the token into the task closure, then inside the closure you can use the token for example to call `FinishOffThreadScript`.

AbhishekSharma102 replied:
Once the `FinishOffThreadScript` has been called, we need to call `evaluate_script_on_global_with_result` with the obtained script, to call `JS_ExecuteScript`. So calling `CompileOffThread` can be called directly from `run_a_classic_script`. Is that correct?

gterzian replied:
I think we need to:

1. Make `run_a_classic_script` take ownership of the `ScriptOrigin`.
2. Store the `ScriptOrigin` on `OffThreadCompilationData`, alongside the `Trusted<Global>`
3. Call the "compile off thread" API.

Then, from the task queued:

1. call `global.evaluate_script_on_global_with_result`.

`run_a_classic_script` doesn't return a result so that should be OK. 

You probably will have to change the signature of `evaluate_script_on_global_with_resul`, making the `code` argument an enum with  a `DomString` and a `JsScript` variant. 

Currently it's a `&str`, but it looks like it could in all current cases be a `DomString` instead.

gterzian replied:
On second thoughts, I think we should do the off-thread compilation earlier, when the data is received from the network. 

So the entry point should I think be around here: https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L268

So I think you can:

Make `ScriptOrigin` have a new `Option<JsScript>` member named `compiled_script`.

1. Store both the `Trusted<HTMLScriptElement>` and the `ScriptOrigin`, and probably also `ExternalScriptKind` onto the `OffThreadCompilationData`.
2. Call the off-thread compilation API
3. From the callback, queue a task that will:
    a. call `FinishOffThreadScript`
    b. Set `compiled_script` to to the result of that call.
    c. continue at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L272

Then `evaluate_script_on_global_with_result` could be changed to take an `Option<JsScript>` as a new `compiled_script` argument, or maybe make the `code` into an enum, and inside you can either call `Evaluate2` or `JS_ExecuteScript`, based on what is available.

AbhishekSharma102 replied:
Just a question, can these scripts be compiled in any context? As GlobalScope is not yet defined and would not be available.

gterzian replied:
> Just a question, can these scripts be compiled in any context?

Yes I think so.

> Also how do we ensure in this that offthreadcompilation is used only when called by run_a_classic_script, as I believe this is done for all scripts?

I think the code at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L268 is only used for classic scripts, if not we can find another way to identify the script and either do the off-thread compilation or simply continue normally.

AbhishekSharma102 replied:
Okay. So I think I will need to create a new task_canceller for this. Is it correct?

gterzian replied:
So I'm almost certain that the code at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L269

is in fact only called for classic scripts, because of the name `ClassicContext` and the spec links, and I also think that is the right point to setup the `OffThreadCompilationData` and then call api to do the off-thread compilation. 

You can get the global at that point from `self.elem.global()`(since `elem` is `HTMLScriptElement`), and then you can get the JS context by doing `global.cx()`. 

Then you also need to store a `Trusted<HTMLScriptElement>` on the `OffThreadCompilationData`, meaning you don't need a `Trusted<GlobalScope>` because you can get the global from the element, and then later in the callback you queue the task that will then continue the algorithm at https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L270

(where again you can get `elem` from the `Trusted<HTMLScriptElement>` you will have available by calling `root()` on it, actually the [`ClassicContext`](https://github.com/servo/servo/blob/1a6193703134512b8d09b54a85bf58b3df82a372/components/script/dom/htmlscriptelement.rs#L186) used there to handle the network responses is very similar to the `OffThreadCompilationData` you will have to define. 

By the way, let's rename it `OffThreadCompilationContext`?

let me know if you have other questions.

gterzian replied:
> o I think I will need to create a new task_canceller for this. Is it correct?

Yes you can call `global.task_canceller()` to get one.

AbhishekSharma102 replied:
I tried getting the globalscope object using `self.elem.global()` but wasn't able to. `HTMLScriptElement` doesn't seem to have it. Can you please confirm?

gterzian replied:
I think you first need to get the root out of the trusted, doing `let elem = self.elem.root(); `, then `elem.global()` should work

AbhishekSharma102 replied:
I had some questions regarding OffThreadCompilation methods we are using.
1. These methods are written in C, but we do an unsafe call to them using a wrapper defined in rust?
2. For JSScript, I get the error that malloc_size_of is not implemented. Upon reading the documentation, I believe it is similar to a string in rust for the time being. So implementing that would mean something to similar done for string?
3. Another error during task queuing is OffThreadToken cannot be passed between threads safely. I think a simple clone should do the job, so that function needs to be implemented for it. Is that correct?

gterzian replied:
1. Yes, I think that's a correct description.
2. You can ignore it, using an annoation like in https://github.com/servo/servo/blob/948dd7379618e85269227a3e37c76a091c85a966/components/script/dom/globalscope.rs#L250
3. I'm not sure about this one. I would say define a `struct OffthreadToken;`, and then in the callback cast the *mut OffThreadToken to it? Then cast it back when you call `FinishOffThreadScript`? @jdm do you have an idea how to represent `OffThreadToken* token` in Rust? It appears to be just an empty class defined at https://github.com/servo/mozjs/blob/aabcc9ba889b2755f1e4e83f28323a60415a790f/mozjs/js/src/vm/HelperThreads.h#L37

gterzian commented:
Could you please add a "FIX #{issue number}" in the opening description for the PR(that way it will be closed when this will, and it also makes it easier to go back to the issue).

Looks like it's going in the right direction!

gterzian commented on components/script/dom/bindings/trace.rs:
```diff
@@ -860,6 +860,12 @@ unsafe impl JSTraceable for StyleLocked<MediaList> {
     }
 }
 
+unsafe impl JSTraceable for JSScript {
```
I think we're going to have to store the script in a `Heap`, and then this won't be necessary, see https://github.com/servo/rust-mozjs/blob/11cabdef2cbf0884a2cc33e3c73719646bd31ce5/src/rust.rs#L465

gterzian commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -53,6 +58,32 @@ use std::sync::{Arc, Mutex};
+) {
+    let off_thread_compilation_context: &mut OffThreadCompilationContext =
+        &mut *(callback_data as *mut OffThreadCompilationContext);
+    let elem = off_thread_compilation_context.script_element.root();
```
this kind of stuff will have to be done from inside the queued task.

gterzian commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -53,6 +58,32 @@ use std::sync::{Arc, Mutex};
+    let _ = off_thread_compilation_context.task_source.queue_with_canceller(
+        task! (off_thread_compile_continue: move || {
+            let global_context = global;
+            off_thread_compilation_context.script_origin.code = SourceCode::Compiled(*FinishOffThreadScript(*global_context.get_cx(), token));
```
So I think we're going to have to do two things:

1. Root the `JsScript` as soon as it is received here, using the `rooted!` macro(you'll see various example across the codebase, just search you'll find many). 
2. store it inside a `Heap` inside the `ScriptOrigin`. For that I think you'll have to do something like:

```rust
let rooted_script = rooted! //
let heap = Heap::default();
heap.set(rooted_script.get());
let source_code = SourceCode::Compiled(heap);
```

AbhishekSharma102 replied:
What is the significance of rooted?

gterzian replied:
https://github.com/servo/servo/blob/6b0d9afd6fdc28356ad44af0104ddd25a7b6438d/components/script/dom/bindings/root.rs#L5

https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/GC_Rooting_Guide

gterzian commented:
Ok so I ended-up looking at this locally, and I think it's actually pretty hard to find the right structure, I ended-up finding something over at https://github.com/gterzian/servo/tree/unblock_off_thread_compilation_pr, which I think could be useful to unblock this PR? 

It took me quite a lot of compilation cycles locally to figure out, so I think it would have been impractical to do with via even more, and probably confusing, review cycles. 

@AbhishekSharma102  How about you rebase this off of that branch? There is still some work to finish, for example in `evaluate_script_on_global_with_result`, and I also suspect there will be some need for further debugging and so on. 

I think there is also an issue that is similar to this and you could use what you've learned to do that one, it is at https://github.com/servo/servo/issues/25209

AbhishekSharma102 commented:
Hey @gterzian ! Thanks for the help. I will take a look. Sorry this is taking so long, as pre occupied with some other work.

AbhishekSharma102 commented:
I keep facing issues in this compilation step.
`rustc --crate-name servo --edition=2018 ports/winit/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type bin --emit=dep-info,link -Cembed-bitcode=no -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="egl"' --cfg 'feature="layout-2013"' --cfg 'feature="max_log_level"' --cfg 'feature="media-gstreamer"' --cfg 'feature="native-bluetooth"' --cfg 'feature="webdriver"' -C metadata=f92413ac43bf4d65 -C extra-filename=-f92413ac43bf4d65 --out-dir /servo_codes/servo/target/debug/deps -C incremental=/servo_codes/servo/target/debug/incremental -L dependency=/servo_codes/servo/target/debug/deps --extern backtrace=/servo_codes/servo/target/debug/deps/libbacktrace-c2de253738f90dda.rlib --extern clipboard=/servo_codes/servo/target/debug/deps/libclipboard-8cdd848607beeec5.rlib --extern euclid=/servo_codes/servo/target/debug/deps/libeuclid-860d44d6f8bac949.rlib --extern getopts=/servo_codes/servo/target/debug/deps/libgetopts-f7c37fe6e97f0526.rlib --extern image=/servo_codes/servo/target/debug/deps/libimage-f5fa0b5fb9d8ed8d.rlib --extern keyboard_types=/servo_codes/servo/target/debug/deps/libkeyboard_types-844dac95fd59d379.rlib --extern lazy_static=/servo_codes/servo/target/debug/deps/liblazy_static-43f70af18d2bceaa.rlib --extern libc=/servo_codes/servo/target/debug/deps/liblibc-61b63b0480794d3d.rlib --extern servo=/servo_codes/servo/target/debug/deps/libservo-ce935d12301a3a40.rlib --extern log=/servo_codes/servo/target/debug/deps/liblog-1483e2d7b9f1cf22.rlib --extern servo_media=/servo_codes/servo/target/debug/deps/libservo_media-bdb2208652d515dc.rlib --extern shellwords=/servo_codes/servo/target/debug/deps/libshellwords-2c1f93ce98cb1ed4.rlib --extern sig=/servo_codes/servo/target/debug/deps/libsig-d7ed8854566421c7.rlib --extern surfman=/servo_codes/servo/target/debug/deps/libsurfman-f4f4a8eb4b1584ae.rlib --extern tinyfiledialogs=/servo_codes/servo/target/debug/deps/libtinyfiledialogs-909cab068201ce77.rlib --extern webxr=/servo_codes/servo/target/debug/deps/libwebxr-c78e8dd8736c4e40.rlib --extern winit=/servo_codes/servo/target/debug/deps/libwinit-3bb6a4b7602c9773.rlib -C link-args=-fuse-ld=gold -W unused-extern-crates -L native=/usr/lib/x86_64-linux-gnu -L native=/usr/lib/x86_64-linux-gnu -L native=/usr/lib/x86_64-linux-gnu -L native=/usr/lib/x86_64-linux-gnu -L native=/usr/lib/x86_64-linux-gnu -L native=/servo_codes/servo/target/debug/build/fontsan-a838246ab2050249/out/lib -L native=/servo_codes/servo/target/debug/build/miniz-sys-f174e987e15b97f4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/servo_codes/servo/target/debug/build/jemalloc-sys-3850958afd320de4/out/build/lib -L native=/usr/lib/x86_64-linux-gnu -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/target/debug/build/mozangle-10b974c9166bcb40/out -L native=/servo_codes/servo/target/debug/build/mozjs-8aa97c24a8623cb4/out -L native=/servo_codes/servo/target/debug/build/mozjs_sys-7554fc71b8fcca7b/out/build/js/src/build -L native=/servo_codes/servo/target/debug/build/mozjs_sys-7554fc71b8fcca7b/out/build/glue -L native=/usr/lib/x86_64-linux-gnu -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/support/linux/gstreamer/gst/lib -L native=/servo_codes/servo/target/debug/build/tinyfiledialogs-4ef66bfd68031559/out`

with this note:
 note: /home/abhishek/.cargo/registry/src/github.com-1ecc6299db9ec823/gstreamer-webrtc-0.15.5/src/auto/enums.rs:274: error: undefined reference to 'gst_webrtc_data_channel_state_get_type'

Any suggestions on this?

gterzian commented:
It might be https://github.com/servo/servo/issues/27090, or another issue related to gstreamer, or it might warrant opening a new issue.

jdm commented:
You will need to run `./mach bootstrap-gstreamer` and uninstall the `gir1.2-gst-plugins-bad-1.0` package.

gterzian commented:
@AbhishekSharma102 Ok thanks for adding the "execute script part", which looks good(although I suspect we'll need to debug it a bit in practice). 

On that last commit, you could add me a co-author, using 2792687+gterzian@users.noreply.github.com as email and following the instructions at https://docs.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

AbhishekSharma102 commented:
@gterzian I will do that. I am planning to squash commits into one, and will add into that. 

Regarding the squashing, unfortunately I merged from master after the first two commits, so there are several comments between the first and last two. Could you suggest on how to merge them?

gterzian commented:
> Regarding the squashing, unfortunately I merged from master after the first two commits, so there are several comments between the first and last two. Could you suggest on how to merge them?

I'm not sure, maybe just try squashing them all into one? 

Usually we don't merge, rather we rebase(and only if there is an actual conflict or it's otherwise necessary), see https://github.com/servo/servo/wiki/Github-workflow

gterzian commented:
I tried rebasing my branch off of servo/master and there were no conflicts, what problems are you encountering?

AbhishekSharma102 commented:
Managed to resolve them. Using git too much for the first time, thus the issues. Merged all commits to one. 

Please let me know if any other changes/tests are needed.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 37358b51e31db3e1909b1b403f689af3d721c883 with merge 30d52366ec2a8974d285527a231a7533b713aa26...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/F0WcdKN3QTmwMqr3HmhqLw)

gterzian commented:
Ok so this clearly broke something, actually it seems to have broken everything, so that's good because it means we just forgot something or made some other obviously mistake, better than having to deal with a few edge cases(might be the next step). 

I'll build locally also and try to help debug.

AbhishekSharma102 commented:
> Ok so this clearly broke something, actually it seems to have broken everything, so that's good because it means we just forgot something or made some other obvious mistake, which is better than having to deal with a few edge cases(might be the next step).
> 
> I'll build locally also and try to help debug.

Any way to obtain more logs?

gterzian commented:
yes via the link above you can look at things like https://community-tc.services.mozilla.com/tasks/b6kS1MK3RbGRj7A4rt1BpA#artifacts

Usually you'd look at the `filtered-intermittents` one, now everything seems to have failed. There are no helpful stack-traces for some reason. I'm hoping to get something better locally .

gterzian commented:
```
pid:840 Stack trace for thread "ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }"
 0:05.02 pid:840    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:05.02 pid:840    1: core::fmt::write
 0:05.02 pid:840    2: std::io::Write::write_fmt
 0:05.02 pid:840    3: servo::backtrace::print
 0:05.02 pid:840    4: servo::install_crash_handler::handler
 0:05.03 pid:840    5: __sigtramp
 0:05.03 pid:840    6: __ZN2JS14CompileOptionsC2EP9JSContext
 0:05.03 pid:840    7: _NewCompileOptions
 0:05.03 pid:840    8: mozjs::rust::CompileOptionsWrapper::new
 0:05.03 pid:840    9: <script::dom::htmlscriptelement::ClassicContext as net_traits::FetchResponseListener>::process_response_eof
 0:05.03 pid:840   10: <net_traits::FetchResponseMsg as net_traits::Action<T>>::process
 0:05.03 pid:840   11: <script::network_listener::ListenerTask<A,Listener> as script::task::TaskOnce>::run_once
 0:05.03 pid:840   12: <T as script::task::TaskBox>::run_box
 0:05.03 pid:840   13: script::script_thread::ScriptThread::handle_msgs
 0:05.03 pid:840   14: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:05.03 pid:840   15: std::sys_common::backtrace::__rust_begin_short_backtrace
 0:05.03 pid:840   16: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:05.03 pid:840   17: std::sys::unix::thread::Thread::new::thread_start
 0:05.03 pid:840   18: __pthread_body
 0:05.04 pid:840   19: __pthread_start
```

Indeed getting a more useful stracktrace locally. 

The failure at `mozjs::rust::CompileOptionsWrapper::new` looks like the classic "we need to enter a realm before calling this"...

I'll give it a try locally.

gterzian commented:
Ok with this change: https://github.com/servo/servo/commit/7b0148f0c409d05a99532a074e0c8298901dfa39

I'm now getting:

```
pid:1563 Stack trace
 0:05.80 pid:1563    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
 0:05.80 pid:1563    1: core::fmt::write
 0:05.80 pid:1563    2: std::io::Write::write_fmt
 0:05.80 pid:1563    3: servo::backtrace::print
 0:05.80 pid:1563    4: servo::install_crash_handler::handler
 0:05.80 pid:1563    5: __sigtramp
 0:05.80 pid:1563    6: script::dom::htmlscriptelement::off_thread_compilation_callback
 0:05.80 pid:1563    7: __ZN2js12HelperThread19handleParseWorkloadERNS_25AutoLockHelperThreadStateE
 0:05.81 pid:1563    8: __ZN2js12HelperThread10threadLoopEv
 0:05.81 pid:1563    9: __ZN2js6detail16ThreadTrampolineIRFvPvEJPNS_12HelperThreadEEE5StartES2_
 0:05.81 pid:1563   10: __pthread_body
 0:05.81 pid:1563   11: __pthread_start
```

jdm commented:
https://github.com/servo/servo/blob/7b0148f0c409d05a99532a074e0c8298901dfa39/components/script/dom/htmlscriptelement.rs#L399-L406 is a stack-owned value and you're passing a pointer to it as the argument to off_thread_compilation_callback. It needs to be a Box::into_raw instead, so that the memory lives long enough to be accessed within the callback.

AbhishekSharma102 commented:
I tried to run it locally, but it gave crash message. Do we need a display to debug? I am running the code in a cloud VM.

gterzian commented:
It seems to work with those changes: https://github.com/servo/servo/commit/7633d85837df02532ade91dafae67c42705550b6 !

@jdm thanks for the pointers(no pun intendend)

@AbhishekSharma102 If you add that commit and squash it into the existing one, we can do another full run and see what happens.

gterzian commented:
> Do we need a display to run wpt? I am running the code in a cloud VM.

I'm not sure.

gterzian commented:
Actually I think I need to make one more change to avoid leaking the token, one moment please.

AbhishekSharma102 commented:
> > Do we need a display to run wpt? I am running the code in a cloud VM.
> 
> I'm not sure.

I'll see if those pass locally (at least some).

gterzian commented:
Ok!

There you go: https://github.com/servo/servo/commit/286a656e4079705584e16d8d2ccae907c8f7a1e0

This one seems to work, and I don't think it leaks.

AbhishekSharma102 commented:
> Ok!
> 
> There you go: [286a656](https://github.com/servo/servo/commit/286a656e4079705584e16d8d2ccae907c8f7a1e0)
> 
> This one seems to work, and I don't think it leaks.

Thanks a lot! Included those changes and pushed.

gterzian commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 24f95fdd926af7d1a00f7f908ce34377d8bfbf7a with merge 302e93205fa238b0d987dabe2608f81e019aa4c8...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/efW-yHZTT4SfhNhrf2EJiQ)

gterzian commented:
These appear relevant:

```
0 matches
2 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /html/semantics/scripting-1/the-script-element/script-onerror-insertion-point-1.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/script-onerror-insertion-point-1.html:
  └ NOTRUN [expected PASS] Test that the insertion point is defined in the error event of a parser-inserted script that actually started a fetch (but just had it fail).
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/load-error-events-1.html:
  │ FAIL [expected PASS] src, 404, parser-inserted, defer, no async
  │   → assert_unreached: Load event should not be fired. Reached unreachable code
  │ FAIL [expected PASS] src, 404, parser-inserted, no defer, no async
  │   → assert_unreached: Load event should not be fired. Reached unreachable code
  │ FAIL [expected PASS] src, 404, parser-inserted, no defer, async
  │   → assert_unreached: Load event should not be fired. Reached unreachable code
  │ 
  │ onLoad/<@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/resources/load-error-events-helpers.js:24:29
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ onLoad@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/resources/load-error-events-helpers.js:16:12
  └ onload@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/load-error-events-1.html:2:7
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/load-error-events-1.html:
  │ FAIL [expected PASS] src, 404, not parser-inserted, no defer, no async, no non-blocking
  │   → assert_unreached: Load event should not be fired. Reached unreachable code
  │ 
  │ onLoad/<@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/resources/load-error-events-helpers.js:24:29
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ onLoad@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/resources/load-error-events-helpers.js:16:12
  └ script3_dynamic_error.onload@http://web-platform.test:8000/html/semantics/scripting-1/the-script-element/load-error-events-1.html:37:44
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/load-error-events-1.html:
  │ FAIL [expected PASS] src, 404, not parser-inserted, no defer, async
```

```
0 matches
1 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  └ libEGL warning: No hardware driver found, falling back to software rendering
  ▶ Unexpected subtest result in /html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html:
  │ TIMEOUT [expected PASS] error event behaves normally (return true does not cancel; one arg) on a script element, with a 404 error
  └   → Test timed out
  ▶ Unexpected subtest result in /html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html:
  └ NOTRUN [expected PASS] error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic Event
  ▶ Unexpected subtest result in /html/webappapis/scripting/events/event-handler-processing-algorithm-error/script-element.html:
  └ NOTRUN [expected PASS] error event behaves normally (return true does not cancel; one arg) on a script element, with a synthetic ErrorEvent

```

Now we've reached the "debug edge conditions" phase...

gterzian commented:
Ok so the problem was not taking `self.status` into account when determining whether the network as an error or not, fixed at https://github.com/servo/servo/commit/8cc91623557cd67e8f45cd09b623e76d2a3f2cc3

AbhishekSharma102 commented:
> Ok so the problem was not taking `self.status` into account when determining whether the network errored, fixed at [8cc9162](https://github.com/servo/servo/commit/8cc91623557cd67e8f45cd09b623e76d2a3f2cc3)

So earlier, in case of an error, we weren't returning, and sending incomplete script to compile off thread?

AbhishekSharma102 commented:
@bors-servo try=wpt

bors-servo commented:
@AbhishekSharma102: :key: Insufficient privileges: not in try users

AbhishekSharma102 commented:
> Ok so the problem was not taking `self.status` into account when determining whether the network errored, fixed at [8cc9162](https://github.com/servo/servo/commit/8cc91623557cd67e8f45cd09b623e76d2a3f2cc3)

Hey, I did the changes, and pushed. I didn't know specific permissions were there for wpt.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 048849bfef031dc9e7c5a6379a67a96b1328ff77 with merge 3c00c5d3f4414f54b97371d5373a31daf2b00593...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/TBWgY4ZzS2-fv3nrv2l9Sg)

gterzian commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -730,7 +893,11 @@ impl HTMLScriptElement {
+            Ok(mut file) => {
+                if let SourceCode::Text(text) = &script.code {
+                    file.write_all(text.as_bytes()).unwrap()
+                }
```
Let's add a `warn!("Cannot store compiled script")` for the `else` case.

gterzian commented:
Getting a PASS for `/custom-elements/customized-built-in-constructor-exceptions.html`, not sure if that is an intermittent. Let's see on the next run.

gterzian commented:
@jdm r? (I'm co-author so don't think I should review)

gterzian commented:
> So earlier, in case of an error, we weren't returning, and sending incomplete script to compile off thread?

Yes

gterzian commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -669,7 +830,9 @@ impl HTMLScriptElement {
-            input.write_all(script.text.as_bytes()).unwrap();
+            if let SourceCode::Text(text) = &script.code {
+                input.write_all(text.as_bytes()).unwrap();
+            }
```
Here I think we should also warn in the else case(or find a way to get the text from a `JSScript`)

AbhishekSharma102 replied:
maybe we can use `JS_DecompileScript `method with uses `context` and `Handle<*mut JSScript>` and returns `*mut JSString`? I am unsure on how to get the context though.

gterzian replied:
Good idea, give it a try I would say. You can get the global from(the equivalent of) `*self.globa().get_cx()`

AbhishekSharma102 replied:
I noticed that the wrapper for  [`DecompileScript`](https://docs.rs/mozjs/0.7.2/mozjs/rust/wrappers/fn.JS_DecompileScript.html) takes 4 args, while the [one](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_reference/JS_DecompileScript) in spidermonkey project, takes 2 and [JS_DecompileScriptObject](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_reference/JS_DecompileScriptObject) takes 4. Due to lack of documentation on the last one, not sure if we can call it.

gterzian replied:
I think you can try to use this one: https://github.com/servo/mozjs/blob/aabcc9ba889b2755f1e4e83f28323a60415a790f/mozjs/js/src/jsapi.h#L1885

gterzian replied:
We could also just store a clone of the original text alongside the compiled script, and use that when needed.

AbhishekSharma102 replied:
Yeah, that's always an option. I think it would be better, as decompile script doesnt always return the original script.

gterzian replied:
yes.

AbhishekSharma102 replied:
So I tried to store it as DOMString, but Send is not implemented for it and cannot queue the task back. Should I implement that similar to how you did for OffThreadToken?

gterzian replied:
I think you should store it as a String then, and later convert it back to a dom string inside the task.

AbhishekSharma102 commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -168,16 +260,43 @@ impl ScriptOrigin {
-        self.text.clone()
+        match &self.code {
+            SourceCode::Text(text) => text.clone(),
+            SourceCode::Compiled(_) => panic!("Cannot convert a compiled script to text."),
```
@gterzian Shall we replace this with the stored script as well? Or is that not needed?

gterzian replied:
Yes we can do it here as well.

AbhishekSharma102 replied:
Updated

gterzian commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -148,9 +231,22 @@ pub enum ScriptType {
+pub struct CompiledSourceCode {
+    #[ignore_malloc_size_of = "SM handles JS values"]
+    pub source_code: Heap<*mut JSScript>,
+    pub script: DOMString,
```
I think it would be clearer if we call this `original_text`.

AbhishekSharma102 replied:
I'll do that change

AbhishekSharma102 replied:
I just pushed the changes to this branch, but not showing here. That's weird

gterzian reviewed (approved):
Looks great, just one nit re the naming

gterzian commented:
@jdm r?

AbhishekSharma102 committed:
Added off thread compilation

Co-authored-by: Gregory Terzian <2792687+gterzian@users.noreply.github.com>
Co-authored-by: Abhishek Sharma <20724848+AbhishekSharma102@users.noreply.github.com>

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -53,6 +63,79 @@ use std::sync::{Arc, Mutex};
+/// A wrapper to mark OffThreadToken as Send,
+/// which should be safe according to
+/// mozjs/js/public/OffThreadScriptCompilation.h
+struct OffThreadCompilationToken(Box<OffThreadToken>);
```
Why do we need a Box?

gterzian replied:
I got a crash at the point of using the token from inside the task that looked very similar to the crash where you pointed out the need to box the pointer, so I think this is a similar situation?

jdm replied:
It's not clear to me how a Box actually helps anything here, since the token is just a pointer. I'd like to see what happens if we remove the Box and just have an OffThreadCompilationToken on the stack that gets used in the task in off_thread_compilation_callback.

Actually, it's not clear to me how the current code even works - we're calling Box::from_raw on the token argument, but we never created it from Box::into_raw so that's invalid.

gterzian replied:
Ok, it's worth a try to see if the boxing is unnecessary @AbhishekSharma102. By the way I think it is valid to use `from_raw` on a pointer that wasn't created via `into_raw`, see the second example at https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw

jdm replied:
You are correct that it can be valid if it was created by a Rust allocator. The OffThreadToken is created by the spidermonkey allocator, though.

gterzian replied:
Ok, thanks for the info.

gterzian replied:
The docs do explain this in more detail over at https://doc.rust-lang.org/std/boxed/index.html#memory-layout (which is linked to from the `from_raw` docs)

AbhishekSharma102 replied:
I am having trouble running wpt locally to test this out. Unit tests are passing without Box as well (i guess that happened last time as well, wpts failed while ut passed)

gterzian replied:
> I am having trouble running wpt locally to test this out.

What problem are you encountering? 

If you don't know exactly which test to run, I think any test will do, since all will load scripts I believe. You could try for example `./mach test-wpt tests/wpt/web-platform-tests/html/`(maybe with the `--release` flag depending on what you have build locally). 

If you don't get a crash, I think it's worth pushing your changes and we can try here as well.

AbhishekSharma102 replied:
I keep getting timeouts, probably due to low sku.

gterzian replied:
yes I think that's normal, but most are ok right? Or does everything time-out? 

Sounds like it's working, and maybe you can address the other comments if not already, and then push the changes and we can take it from there?

AbhishekSharma102 replied:
Apart from CompileOffThread1, I've addressed the rest.

 Will push after rebasing.

AbhishekSharma102 replied:
when i do ./mach build --dev, it tries to update a few things and demands for almost 52GB of memory, and fails.
memory allocation of 52834435104 bytes failed.

Isnt that a bit too much?

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -53,6 +63,79 @@ use std::sync::{Arc, Mutex};
+    let final_url = context.final_url.clone();
+    let script_element = context.script_element.clone();
+    let script_kind = context.script_kind.clone();
+    let script = context.script_text.clone();
```
I'm concerned about cloning the string text here, because these could be very large minified scripts. What if we used `mem::replace(context.script_text, String::new())` instead?

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -148,9 +231,22 @@ pub enum ScriptType {
 }
 
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct CompiledSourceCode {
```
We either need to add a `#[must_root]` to this struct (and store it within a `RootedTraceableBox`) since it contains a GC value, or we need to store `pub source_code: RootedTraceableBox<Heap<*mut JSScript>`.

gterzian replied:
I think we are only passing this around, so then I guess using the `RootedTraceableBox` is the right option? I thougth a `Heap` would also root the value, however I guess that is not correct, and most use of `Heap` is within a dom struct which roots it...

jdm replied:
Storing the RootedTraceableBox as a member of CompiledSourceCode should be the easiest solution.

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -243,42 +366,85 @@ impl FetchResponseListener for ClassicContext {
-        let load = response.and(self.status.clone()).map(|_| {
-            let metadata = self.metadata.take().unwrap();
+        let (source_text, final_url) = match (response, self.status.as_ref()) {
+            (Err(err), _) => {
+                // Step 6, response is an error.
+                finish_fetching_a_classic_script(
+                    &*self.elem.root(),
+                    self.kind.clone(),
+                    self.url.clone(),
+                    Err(err),
+                );
+                return;
+            },
+            (_, Err(err)) => {
+                // Step 6, response is an error.
+                finish_fetching_a_classic_script(
+                    &*self.elem.root(),
+                    self.kind.clone(),
+                    self.url.clone(),
+                    Err(err.clone()),
+                );
+                return;
+            },
```
Could we combine these as `(Err(err), _) | (_, Err(err))` and avoid the duplication?

gterzian replied:
There was an issue because the `_` is not the same type in those two arms(there might be another way to make it work, but that is why I added two different match arms).

jdm replied:
Are you sure it's the _ that matters? What if you matched on `response.as_ref()`?

gterzian replied:
I'm not sure, I'll leave it to @AbhishekSharma102 to look into.

AbhishekSharma102 replied:
Using `response.as_ref()` doesn't throw any errors. Have merged the two

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -243,42 +366,85 @@ impl FetchResponseListener for ClassicContext {
+                script_kind: self.kind.clone(),
+                final_url,
+                url: self.url.clone(),
+                task_source: global.networking_task_source(),
+                canceller: global.task_canceller(TaskSourceName::Networking),
```
Let's use the dom_task_source instead.

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -243,42 +366,85 @@ impl FetchResponseListener for ClassicContext {
+            let source_text: Vec<u16> = source_text.encode_utf16().collect();
+
+            unsafe {
+                assert!(CompileOffThread(
```
If we use CompileOffThread1 we can pass a UTF8 string instead. We could even store the source text inside of an Arc and avoid having to clone the underlying source at all.

However, this could also be a followup issue.

AbhishekSharma102 replied:
I think we can use this straightaway, without changing too much code. I'll try to see what can be done.

jdm reviewed (changes requested):
I found this straightforward to read. Great work, both of you!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #27282) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -48,13 +59,89 @@ use servo_url::ServoUrl;
+                    code,
+                    url: final_url,
+                    external: true,
+                    fetch_options: ScriptFetchOptions::default_classic_script(&global),
```
Let's store the ScriptFetchOptions in OffThreadCompilationContext and use it here instead.

jdm commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -261,43 +388,76 @@ impl FetchResponseListener for ClassicContext {
-                metadata.final_url,
-                self.fetch_options.clone(),
+                final_url.clone(),
+                ScriptFetchOptions::default_classic_script(&global),
```
We should be using `self.fetch_options.clone()` here still.

AbhishekSharma102 replied:
should we also store self.fetch_options in OffThreadCompilationContext?

jdm reviewed (changes requested):
This is very close!

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 2d9f1446af7b55b288fb618d56f50a999d9c3146 with merge 4db6fd9e415c5323569a9ed4ace48ab69a5dc40e...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/FdFlupJcTKW-88UZSmAfIg)

AbhishekSharma102 commented on components/script/dom/htmlscriptelement.rs:
```diff
@@ -48,13 +59,89 @@ use servo_url::ServoUrl;
+                let script_text = DOMString::from(script);
+                let heap = Heap::default();
+                heap.set(compiled_script.get());
+                let source_code = RootedTraceableBox::new(heap);
```
Can you please explain a bit on this or share some doc?

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 24d80c22b0bcfdb926073a60f582cb8cbeabf149 with merge 503e6037566714c4b0551c6c2a91400c8a903fc2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZAXJi_PpQ4KEn0cGfkWDnA)

jdm commented:
```
  ▶ TIMEOUT [expected OK] /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  └ [2020-07-20T19:57:55Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html:7:17 requestViaXhr is not defined
  ▶ Unexpected subtest result in /referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html:
  │ TIMEOUT [expected FAIL] Sandboxed iframe with opaque origin doesn't send referrers.
  └   → Test timed out
  ▶ OK [expected CRASH] /custom-elements/customized-built-in-constructor-exceptions.html

  ▶ ERROR [expected OK] /_webgl/conformance/more/functions/bindFramebufferLeaveNonZero.html
  │   → Tests is not defined
  │ 
  └ http://web-platform.test:8000/_webgl/conformance/more/functions/bindFramebufferLeaveNonZero.html:40:1
```
Very interesting. I'd like to take a closer look at those failures to understand what's going on.

gterzian commented:
> requestViaXhr is not defined

That test is loading a script via the srcdoc of an iframe. https://github.com/servo/servo/blob/1ded1ba9926f962bd84e9c949677009da06ea653/tests/wpt/web-platform-tests/referrer-policy/generic/sandboxed-iframe-with-opaque-origin.html#L28

I'm guessing the fact that we're not blocking on compiling/executing the script when it loads exposes some sort of timing issue with running the code of an srcdoc?

gterzian commented:
So when an iframe has an srcdoc, we parse and execute it over at https://github.com/servo/servo/blob/1ded1ba9926f962bd84e9c949677009da06ea653/components/script/script_thread.rs#L3904

I think it's  a problem with not blocking the load event properly on that src doc while we're still compiling the script? This must have been hidden previously due to the sync execution?

gterzian commented:
Or maybe we should apply the off-thread compilation flow also to classic scripts with no src(as used in that srcdoc), see https://github.com/servo/servo/blob/1ded1ba9926f962bd84e9c949677009da06ea653/components/script/dom/htmlscriptelement.rs#L640

jdm commented:
The way I understand it is that since we only try to compile external scripts off the main thread, the parser should already be blocked and therefore we shouldn't parse the internal script before we're finished executing the external scripts. If that's not the case, we need to figure out why the parser isn't blocked.

jdm commented:
The parser is indeed blocked as expected. It turns out that off-thread compilation is failing for one instance of common.sub.js in the srcdoc iframe, while the other succeeds. RUST_LOG=script::dom::htmlscriptelement demonstrates this happening.

jdm commented:
Aha - we're violating the rule that [script source characters need to live until the callback is called](https://searchfox.org/mozilla-central/rev/3b6958c26049c1e27b2790a43154caaba9f6dd4a/js/public/OffThreadScriptCompilation.h#55-57), so we're racing with the off-thread parser as we destroy the utf-16 characters that we pass when we call CompileOffThread.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 69488c13d1ac1effb1a61c1120b6c94ea1cc7a00 with merge 5fd9598cc8b56ea3af0a391cc89501dee7d1bfac...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/CX194KAURIqh_3_p46lMGw)

jdm commented:
@bors-servo r=jdm,gterzian

bors-servo commented:
:pushpin: Commit 1119dd1 has been approved by `jdm,gterzian`

<!-- @bors-servo r=jdm,gterzian 1119dd119ed335b6d6e448290b56e0eeab6de25f -->

AbhishekSharma102 commented:
> Aha - we're violating the rule that [script source characters need to live until the callback is called](https://searchfox.org/mozilla-central/rev/3b6958c26049c1e27b2790a43154caaba9f6dd4a/js/public/OffThreadScriptCompilation.h#55-57), so we're racing with the off-thread parser as we destroy the utf-16 characters that we pass when we call CompileOffThread.

I seem to have completely missed this. Sorry for this.

bors-servo commented:
:hourglass: Testing commit 1119dd119ed335b6d6e448290b56e0eeab6de25f with merge 82a9e9d29259ee76b49817c62ca6100072c67f4f...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/BQK6CJtsT0KK3BUtb2zFtw)

jdm commented:
@bors-servo retry
* #25513 #27342

bors-servo commented:
:hourglass: Testing commit 1119dd119ed335b6d6e448290b56e0eeab6de25f with merge a24291362946f92bf155b390ea469b7be50a4c2d...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/TcdxVMybRoaX70ERRtv0LA)
Approved by: jdm,gterzian
Pushing a24291362946f92bf155b390ea469b7be50a4c2d to master...

gterzian commented:
@jdm thanks for catching that. You added the tracer right? So you mean previously it could be GC'ed during the off-thread compilation?

jdm commented:
Not GCed; we called `source_text.encode_utf16().collect()` before passing the resulting vector's pointer to CompileOffThread, but then we would drop the vector when we left that stack frame. I moved the vector into the OffThreadCompileContext to ensure that it remained alive until we dropped that context inside the off thread compilation callback function.

jdm commented:
I don't think I added the JSScript tracer to this PR. I guess we just haven't ever tried to store a JSScript in a Heap before?

gterzian commented:
> I moved the vector into the OffThreadCompileContext to ensure that it remained alive until we dropped that context inside the off thread compilation callback function.

Makes sense, thanks for the info. 

>  I guess we just haven't ever tried to store a JSScript in a Heap before?

I think so too.

There is this: https://github.com/servo/rust-mozjs/blob/0caf5549cddbb34ad16abf35fb6bfbff824a4d14/src/rust.rs#L466

However that's `Trace`, not `JsTraceable`...

AbhishekSharma102 commented:
> I don't think I added the JSScript tracer to this PR. I guess we just haven't ever tried to store a JSScript in a Heap before?

I added the tracer as RootedTraceableBox needed it i think. So did something similar to that was done for JSString and JSObject. Hope that was correct.

Also, can you suggest a easy/medium issue that I could work on next? Should i pick up #25209 ?

gterzian commented:
https://github.com/servo/servo/issues/25209 is a good one I think, yes.
